### PR TITLE
fix: allow optional remote form schema fields under `{exactOptionalPropertyTypes: true}`

### LINKED ISSUE #15568: `exactOptionalPropertyTypes` does not play well with optional props in remote form input schemas
### Describe the bug

Enabling
```
{
	"compilerOptions": { "exactOptionalPropertyTypes": true }
}
```
in `tsconfig.json` causes type errors when using optional fields in the schema of a remote `form()` function

consequently, the type of the parsed data is not correctly inferred, and `svelte-check` will error with exit code `1`

### Reproduction

```ts
import * as v from 'valibot';
import { form } from '$app/server';

export const createPost = form(
	v.object({
		// this seems to work fine, but TS and `svelte-check` will complain
		myField: v.optional(v.string()),
	}),
	async ({ myField }) => {
		/* ... */
	},
)
```

### Logs

```Shell
│ /src/routes/app/example.remote.ts:5:2
│ Error: No overload matches this call.
│   Overload 1 of 3, '(validate: "unchecked", fn: (data: RemoteFormInput, issue: Record<string | number, any>) =>…
│     Argument of type 'ObjectSchema<{ readonly myField: OptionalSchema<StringSchema<undefined>, undefined>; }, u…
│   Overload 2 of 3, '(validate: StandardSchemaV1<RemoteFormInput, Record<string, any>>, fn: (data: Record<string…
│     Argument of type 'ObjectSchema<{ readonly myField: OptionalSchema<StringSchema<undefined>, undefined>; }, u…
│       The types of '"~standard".types' are incompatible between these types.
│         Type 'StandardTypes<{ myField?: string | undefined; }, { myField?: string | undefined; }> | undefined' …
│           Type 'StandardTypes<{ myField?: string | undefined; }, { myField?: string | undefined; }>' is not ass…
│             Types of property 'input' are incompatible.
│               Type '{ myField?: string | undefined; }' is not assignable to type 'RemoteFormInput'.
│                 Property 'myField' is incompatible with index signature.
│                   Type 'string | undefined' is not assignable to type 'MaybeArray<string | number | boolean | R…
│                     Type 'undefined' is not assignable to type 'MaybeArray<string | number | boolean | RemoteFo…
│ export const signIn = form(
│ 	v.object({
│ 		myField: v.optional(v.string()),
│ 	}),
│ 	async ({myField}) => {
```

### System Info

```Shell
Binaries:
    Node: 24.11.1
    npm: 11.6.2
    pnpm: 10.17.1
  npmPackages:
    @sveltejs/adapter-cloudflare: ^7.2.6 => 7.2.8
    @sveltejs/kit: ^2.55.0 => 2.55.0
    @sveltejs/vite-plugin-svelte: ^7.0.0 => 7.0.0
    svelte: ^5.54.0 => 5.54.0
    vite: ^8.0.1 => 8.0.1
```

### Severity

serious, but I can work around it

### Additional Information

_No response_

### PR DESCRIPTION
**Pull request by echocrow**
closes #15568

Naively allows `undefined` type in form input values to satisfy type checks when `exactOptionalPropertyTypes` is enabled.

This PR also adds previously-failing tests. To actually test for this, the TS compiler option `exactOptionalPropertyTypes: true` must be set.
Seems that the current testing strategy for the `test/types` file was to run `tsc`. Unfortunately this throws all sorts of errors when with `exactOptionalPropertyTypes` for the new file, because `tsc` doesn't limit that compiler option to just that one test file, but also to all other included/imported files, too. However, kit core was not written with that compiler option in mind, resulting in a lot of false negatives.
To bypass that issue and still add tests, I opted to move the type check automation from `tsc` to `vitest typecheck`. Technically Vitest's typecheck is still experimental, but we've been using it successfully for some time.

---

### Please don't delete this checklist! Before submitting the PR, please make sure you do the following:
- [x] It's really useful if your PR references an issue where it is discussed ahead of time. In many cases, features are absent for a reason. For large changes, please create an RFC: https://github.com/sveltejs/rfcs
- [x] This message body should clearly illustrate what problems it solves.
- [x] Ideally, include a test that fails without this PR but passes with it.

### Tests
- [x] Run the tests with `pnpm test` and lint the project with `pnpm lint` and `pnpm check`

### Changesets
- [x] If your PR makes a change that should be noted in one or more packages' changelogs, generate a changeset by running `pnpm changeset` and following the prompts. Changesets that add features should be `minor` and those that fix bugs should be `patch`. Please prefix changeset messages with `feat:`, `fix:`, or `chore:`.

### Edits

- [x] Please ensure that 'Allow edits from maintainers' is checked. PRs without this option may be closed.

### CODE REVIEW

File: packages/kit/test/types/vitest.config.ts
Comment (echocrow) [inline packages/kit/test/types/vitest.config.ts:1]: I noticed that kit's main Vitest config was given a custom name along with this note:
```
// this file needs a custom name so that the numerous test subprojects don't all pick it up
```

does the requirement apply here too? if so, could name it e.g. `kit-types.vitest.config.js`. either way, may make sense to co-locate it next to `kit.vitest.config.js`(?)

_alternatively_, we could probably also move this config into `kit.vitest.config.js`. by default, that would also move type checks out of the `check` command and become part of `test:unit` instead. could maybe avoid that by filtering tests, but separate configs are probably preferred, if type checks and unit tests should be associated with different scripts

Comment (teemingc) [reply packages/kit/test/types/vitest.config.ts:1]: I'd recommend not using Vitest for this at all. Our `pnpm check` command simply navigates to the folder with the `tsconfig.json` file and runs `tsc` to see if it spits out type compilation errors. You could extend the npm script to navigate into another folder with a different `tsconfig.json`

**Review by teemingc — DISMISSED** [review]

File: packages/kit/test/types/vitest.config.ts
Comment (echocrow) [reply packages/kit/test/types/vitest.config.ts:1]: > you could extend the npm script to navigate into another folder with a different tsconfig.json

@teemingc that's exactly what I'd tried at first!

the `check` command got a little bloated, but that was the lesser issue. the bigger issue with that approach is that it spits out a wall of additional compile errors for the added `exactOptionalPropertyTypes` tests, because `tsc` would apply `{exactOptionalPropertyTypes: true}` to kit source code imported in the test files too, essentially type checking parts of `kit` src with that compiler flag. but that source code was not written with that flag in mind, hence a spam of false positives.

as far as i can tell, `vitest typecheck` is effectively running type compile checks via `tsc` (unless another [`checker`](https://vitest.dev/config/typecheck.html#typecheck-checker) was explicitly configured). by default, that would cause the same false-positive type errors as just running `tsc` directly. however, vitest implemented a convenient `ignoreSourceErrors` flag (enabled in this PR) that only flags tsc errors that were reported in the test code, but ones that originated from the source code

also noteworthy: i did _not_ remove the main `tsc` for type checking kit's source code! i've only moved the type checks for `test/types` to `vitest typecheck`

so the options i'm seeing:
- (a) vitest typecheck with `ignoreSourceErrors` to ignore irrelevant source code
  - that's what this PR is doing
- (b) yolo and not add type checks for the `exactOptionalPropertyTypes` case
- (c) keep tsc, but somehow ignore source code errors (aka what vitest implemented)?
- (d) alter a bunch of types in `kit` source with `exactOptionalPropertyTypes` in mind
  - for the most part, migrating a bunch of `{myOptionalKey? MyType}` tpo `{myOptionalKey? MyType | undefined}`. this is a niche opinionated choice, and probably not desired
  - this would need to be done either for all types, or "just the types that coincidentally get checked by tsc from `remote.test.ts`"

(a) seemed like the least impactful choice while still adding a previously-breaking test

Comment (teemingc) [reply packages/kit/test/types/vitest.config.ts:1]: Thanks for explaining that. I wonder if this means that users who run `tsc` will run into this issue too, effectively making this a band aid rather than a proper fix. Maybe supporting `exactOptionalPropertyTypes` requires a bigger change to the Kit codebase

Comment (echocrow) [reply packages/kit/test/types/vitest.config.ts:1]: (edit: typed this before i saw your new post above, following up on that in another post)

update: took a 2nd stab at this via `tsc` directly

because kit src code is written in JS, there is an option to replicate vitest typecheck's `ignoreSourceErrors` with these compilerOptions for the added `test/type`'s config
```jsonc
"checkJs": false,
"skipLibCheck": true
```
this then effectively ignores tsc errors in kit src code when running `tsc` on e.g. the newly added `exactOptionalPropertyTypes/remote.test.ts`

this is more of a coincidental property, rather than vitests explicit and more descriptive `ignoreSourceErrors`

@teemingc with all that in mind, do you still prefer keeping the previous `tsc` approach, or you cool with the suggested move to`vitest typecheck` for just kit's `test/types`? (`tsc` for kit `src` is kept as-is and vitest still uses `tsc` for typecheck under the hood)

Comment (echocrow) [reply packages/kit/test/types/vitest.config.ts:1]: > I wonder if this means that users who run tsc will run into this issue too

as in users to run `tsc` on `kit/src` with `exactOptionalPropertyTypes`? I may be missing something, but I wouldnt deem that a problem

outside of this repo, I believe that _can_ throw type errors when running `tsc` with a variety/combination of compiler options _other than_ what the package's source code was authored in, not just with `exactOptionalPropertyTypes`. but this is a general issue of the node/npm/typescript packaging ecosystem as a whole. I believe the common workaround is to enable `"skipLibCheck": true` in the user's tsconfig.

(that is also exactly why i had to include `"skipLibCheck": true` in the `tsc`-only approach mentioned above. otherwise I was getting a singular random type error under `exactOptionalPropertyTypes` from an imported `../../../../../node_modules/{...}/vite/dist/node/index.d.ts` file. 💀)

> Maybe supporting exactOptionalPropertyTypes requires a bigger change to the Kit codebase

that would also be an option! and would eliminate the need to an extra type check setup. i personally enable `exactOptionalPropertyTypes` more often than not, but I understand it's an opinionated choice, and one that sometimes comes with a little extra headache. happy to assist with such a migration, but would like to get prior approval from the maintainers first. it's a niche compiler option that I suspect most people don't care for.

Comment (teemingc) [reply packages/kit/test/types/vitest.config.ts:1]: I think historically we’ve favoured the types working without skipping any checks such as `skipLibCheck` as we have users who don’t enable it

Comment (echocrow) [reply packages/kit/test/types/vitest.config.ts:1]: in that case, i guess the choice is between
- (a) not supporting `exactOptionalPropertyTypes` (probably leave as-is & close PR)
- (b) adopt `exactOptionalPropertyTypes` in kit

(b) quietly opens up the question of which other niche compilerOptions to support or not support, but maybe it's fine on a case-by-case basis(?) (and I'm saying this as someone typically in favor of `exactOptionalPropertyTypes`)

outside of those, currently this PR sets up one way for granular type tests under a variety of different compiler options, without requiring kit src to support them all. (but, as you said, at the expense of potentially `tsc` erroring in user land with colliding user-specified compilerOptions and without `skipLibCheck`)

Comment (echocrow) [reply packages/kit/test/types/vitest.config.ts:1]: for what it's worth, im not getting any type errors when running `tsc` in a project that imports a couple features from `$app/*` and `@svelte/kit` while forcing `"exactOptionalPropertyTypes": true, "skipLibCheck": false,`

but to be candid, I also dont confidently understand why I'm _not_ getting `tsc` errors in userland with those compilerOptions. my guess is that the explicit `types/index.d.ts`/`ambient.d.ts` is effectively causing tsc to type-check only those explicit types, skipping the internal implementations of kit that _would_ error under `exactOptionalPropertyTypes`

so unless i'm missing something, perhaps `tsc` concern mentioned [here](https://github.com/sveltejs/kit/pull/15866#discussion_r3265536808) is not an issue? (neither before nor after this PR)

Comment (teemingc) [reply packages/kit/test/types/vitest.config.ts:1]: I think if we do decide to support `exactOptionalPropertyTypes`, we'd just enable it for the whole codebase to ensure there's full compatibility for it. Otherwise, we could still be emitting errors from other packages than kit or remote function types.

Comment (echocrow) [reply packages/kit/test/types/vitest.config.ts:1]: makes sense. if/when that comes, happy to help with the migration!

where does this leave this PR for now? we'd been using svelte/sveltekit for a while, most of that with `exactOptionalPropertyTypes` enabled. the new remote `form()` input schema is the only instance i'd encountered so far that threw a false type error under `exactOptionalPropertyTypes` only.

for now, this PR addresses just that one schema type, and adds a basic type check for it. PR rn still switches `test/types` to vitest typecheck (while keeping `tsc` check for `kit/src`), but happy to revert `test/types`  to tsc directly. - or i could also nuke the added tests and change test setup altogether, only keeping the added `| undefined` for now. 

outside of those, guessing #15568 will be a wontfix? and i'll just need to close my eyes when using optional form fields (🥲)

Comment (teemingc) [reply packages/kit/test/types/vitest.config.ts:1]: I think we can probably merge this in without the test since it's more setup than it's worth at the moment. I'll let another maintainer who's more familiar with remote functions look at this

Comment (dummdidumm) [reply packages/kit/test/types/vitest.config.ts:1]: I agree, let's revert the vitest stuff, if there's no "optional property" test then add one (but without the tsconfig adjustment) so we can at least check that the change does not break the current configs.

Comment (echocrow) [reply packages/kit/test/types/vitest.config.ts:1]: @dummdidumm cheers for hopping on!

> if there's no "optional property" test then add one

there is; see [this diff](https://github.com/sveltejs/kit/pull/15866/changes#diff-26795f211df80afb87b3c57488f9ef45f054b3f06cef25691153ff5b0576b698)

> (but without the tsconfig adjustment)

how do you propose we should go about this?

from what i've seen and tested, the tsconfig for this edge case is needed to get a failing test before the change. without it, the test would have not failed, and thus doesnt add value(?). per linked issue, the type error does not occur without `exactOptionalPropertyTypes`; but with that compiler option, the example in the docs will throw a type error (see repro in #15568)

summarizing from earlier comments, options im seeing (from low to high effort):
- (a) no change, wont fix, boo me
- (b) just ship the `| undefined` type addition to `RemoteFormInput`, hope it wont break in the future
- (c) add type addition & a test case (with dedicated tsconfig with `{exactOptionalPropertyTypes: true}`); test with `tsc` (needs some more tsconfig juggling to avoid false negatives, see [this comment](https://github.com/sveltejs/kit/pull/15866#discussion_r3265572371))
- (d) add type addition & a test case (with dedicated tsconfig with `{exactOptionalPropertyTypes: true}`); test with `vitest typecheck` (uses `tsc` under the hood; with `ignoreSourceErrors` to avoid false positives; see [this comment](https://github.com/sveltejs/kit/pull/15866#discussion_r3265462251) for details) - current state of this PR
- (e) fully adopt `exactOptionalPropertyTypes` in kit, fix new kit src type errors as needed (mostly adding `| undefined` to a bunch of optional properties), add the test case (this time without a dedicated tsconfig; base tsconfig now covers `exactOptionalPropertyTypes`)

not sure if there is an option between (b) and (c) as you're asking? but maybe i'm missing something

personally leaning (d), but will revert if vitest typecheck is not welcome for checking types. just need to understand how you'd like me to handle the type test for this compilerOption-specific case

Comment (echocrow) [reply packages/kit/test/types/vitest.config.ts:1]: just saw that #15825 got merged (🎉!) which addresses another quirk in projects with `exactOptionalPropertyTypes`. looks like it tests the behavior by making `exactOptionalPropertyTypes` part of the tsconfig already in place for other related tests (in `packages/kit/src/core/sync/write_types/test/app-types/tsconfig.json`)

if that's an acceptable change, i can roll with the same approach here and drop the newly added tsconfig.
TBD how to go about the type errors in source code under `exactOptionalPropertyTypes`. i suspect the overridden `paths` for `@sveltejs/kit` and `$app/types` is how that `app-types` test  bypasses this issue; fingers crossed the same idea works and is accepted here

Comment (echocrow) [reply packages/kit/test/types/vitest.config.ts:1]: latest version is up! @dummdidumm 

back to a single `tsconfig`, and the `tsc` approach for type checking as requested

inconveniences along the way:
- I still had to sneak in a `skipLibCheck`
  - couldn't figure out another workaround to silence a tsc error in a vite dependency.
  - it seems that the tests in `write_types` use the same workaround; search for `--skipLibCheck`.
- couldn't find a workaround for some source type errors under `exactOptionalPropertyTypes`
  - I believe `write_types` doesn't run into these, because, unlike these tests here, it does not import from `$app/server`
  - I opted to partially patch the types were necessary for the types to pass. that did, unfortunately, result in some awkward type casts though.

Comment (echocrow) [reply packages/kit/test/types/vitest.config.ts:1]: @teemingc sry for the bump; hope you dont mind me re-requesting review(?). guessing dummdidumm is busy or the notif was dropped.

PR should hopefully align with [their request](https://github.com/sveltejs/kit/pull/15866#discussion_r3272345213), but to accommodate I had to resolve some type issues in kit src under `exactOptionalPropertyTypes`

lmk if you'd like me to rebase first (in case any of the recent kit changes introduced new type errors in the type check)

File: packages/kit/src/runtime/client/client.js [hunk 1]
```diff
@@ -971,7 +978,10 @@ async function load_node({ loader, parent, url, params, route, server_data_node
 						// To keep the two values in sync, we explicitly set the headers to `undefined`.
 						// Also, not sure why, but sometimes 0 is evaluated as truthy so we need to
 						// explicitly compare the headers length to a number here
-						headers: [...resource.headers].length > 0 ? resource?.headers : undefined,
+						headers:
+							[...resource.headers].length > 0
+								? resource?.headers
+								: /** @type {never} */ (undefined),
```
Comment (elliott-with-the-longest-name-on-github) [inline packages/kit/src/runtime/client/client.js]: Why do we have these `/** @type {never} */` casts everywhere? It seems a weird way to just lie about what we're doing

**Review by elliott-with-the-longest-name-on-github — COMMENTED** [review]
This seems good, but I'm weirded out by the casts to `never` -- why are they necessary?

File: packages/kit/src/runtime/client/client.js (hunk 1 above)
Comment (echocrow) [reply packages/kit/src/runtime/client/client.js]: yeah NGL i hated myself when i typed that. but couldnt ad-hoc figure out a setup that didnt require a bit more refactor or a change to runtime code

IIRC:
- `RequestInit['headers']` is optional, but not typed as `| undefined`
- at runtime, `{headers: undefined}` is fine; will be treated the same way as `{}`
- however, under `exactOptionalPropertyTypes`, passing `undefined` causes typescript to complain during type check:
    ```
    Type 'undefined' is not assignable to type 'HeadersInit'.
    ```

some alternative options
- pass `[]` instead of `undefined` as "else" value
  - probably the simplest solution; just a small runetime change, but one that should not have any impact
  - hadn't done so yet; wanted to avoid runtime changes for the purpose of this type-specific PR if possible
  - in that case could maybe further simplified to just `headers: [...resource.headers]`
- progressively build a `requestInit` object
  - a la `if (headers.length) myRquestInit.headers = headers`, then pass or spread that into the 2nd fetch arg
- patch `RequestInit['header']` from `HeadersInit` to `HeadersInit | undefined`
  - ...but messing with DOM type defs seems even more fishy

more workarounds are probably available

alternatively, one of the earlier iterations had a slightly different setup that avoided typechecking src code under `exactOptionalPropertyTypes`. in that case, we don't need to make src pass under `exactOptionalPropertyTypes`. that required either a separate tsconfig though, or a wrapper like vitest to ignore source code errors

alternatively-alternatively, give up & don't support `exactOptionalPropertyTypes`. it's quirks like this with the DOM lib that i'd been considering dropping this tsconfig flag for our repos more and more

Comment (elliott-with-the-longest-name-on-github) [reply packages/kit/src/runtime/client/client.js]: I do think `exactOptionalPropertyTypes` falls under the same umbrella of `noUncheckedIndexedAccess` -- theoretically useful but unfortunately TypeScript and the ecosystem is completely unprepared to deal with the horrible edge cases you run into... 

I'm a bit leery of anything that requires us to play by `exactOptionalPropertyTypes` rules in the SvelteKit codebase as it's more-often-than-not more of an esoteric thing than an actually useful signal. Maybe the most pragmatic thing we can do here is fix the external API issue but not add tests for it? The downside is that obviously we could break this in the future for users of `exactOptionalPropertyTypes` (there are dozens of us)!

Comment (echocrow) [reply packages/kit/src/runtime/client/client.js]: > I'm a bit leery of anything that requires us to play by `exactOptionalPropertyTypes` rules in the SvelteKit codebase 

+1

> Maybe the most pragmatic thing we can do here is fix the external API issue but not add tests for it?

my personal choice would be to switch to vitest for the type test, with the `ignoreSourceErrors` config enabled. this allows us to add a failing test with specific tsconfig flags that'll apply to the APIs used, without requiring source code to also conform those tsconfig options. have the cake and eat it too! but i also get wanting to keep external deps low and stick with the `tsc` cli directly.

short of that, also cool with undoing all changes source and the failing test, effectively only fixing the API type. seems low-chance/low-risk, and quick to re-fix should this ever re-occur. (I believe this is [also the conclusion](https://github.com/sveltejs/kit/pull/15866#discussion_r3267820294) teemingc landed at)

short of both, also fine to close this as wontfix - one more pin in `exactOptionalPropertyTypes`'s coffin.

> (there are dozens of us)

probably rounded up 😅

Comment (elliott-with-the-longest-name-on-github) [reply packages/kit/src/runtime/client/client.js]: Yeah, let's go with "just fix the issue and if it ever breaks again someone can holler". If it becomes a frequent issue in the future we can totally add something like this testing setup back in, but I'm somewhat doubtful that it will

**Review by teemingc — CHANGES_REQUESTED** [review]
See https://github.com/sveltejs/kit/pull/15866/changes#r3431761205

**Comment by echocrow** [comment]
type-only fix pushed as per [this comment](https://github.com/sveltejs/kit/pull/15866#discussion_r3431761205) - re-requesting review

**Review by teemingc — APPROVED** [review]
Thanks!
